Add new `tinyint` and `smallint` abstract schema types to support smaller storage requirements

sebastienbarre opened the pull request:
Per preliminary discussion in #10336 this PR:
* adds new `tinyint` and `smallint` abstract schema types to support smaller storage requirements.
* ~~adds an optional `storage` field to the schema abstraction so that various flavors of integers can be supported by specific datasources. This field provides a hint as to how many bytes are required to efficiently store the contents of a column. Backend can choose to provide and honor such hint.~~

In other words, this allows for a MySQL backend to keep its `smallint(6)`, `tinyint(4)`, or `tinyint(2)` intact instead of having them turn into `int(6)`, `int(4)`, or `int(2)`. This is particularly important when a table stores a million+ rows and a column can be stored as one byte (`tinyint`) instead of 4 (`int`).

Tests are provided.

sebastienbarre committed:
allow data sources to provide smaller integers based on storage requirements

sebastienbarre committed:
fix coding standard

markstory commented on lib/Cake/Model/Datasource/DboSource.php:
```diff
@@ -3459,12 +3469,16 @@ public function buildColumn($column) {
-			return null;
+		// if a storage requirement in bytes was set, try it first (i.e. 'integer/4' before 'integer')
+		if (isset($storage) && isset($this->columns[$type . '/' . $storage])) {
+			$real = $this->columns[$type . '/' . $storage];
```
Why would storage lengths be embedded in the column name? I saw a 'storage' option elsewhere in these changes. A specific option seems like a better solution to me.

sebastienbarre replied:
Because there is unfortunately no API in `DboSource::buildColumn($column)` that given *all* the fields in `$column` that describe the column in an abstract way (i.e. `type` *and* the new `storage` field) would return the proper real/native type like `smallint`. There *is* such method when going the other way around -- `DataSource::column()` takes a real column type and return an abstract type.

Instead, `buildColumn()` is *directly* poking at `$this->columns[$type]`, an *array* that apparently all datasources subclasses are supposed to have in place. What this means to me is that I couldn't introduce a new API like `DboSource::abstractToRealColumn()` (the opposite of `column()`) which would take the whole `$column` array of fields as a parameter instead of just `$type`, because if people have implemented their own data sources with a `$this->columns[]` array as they are supposed to, then these will no longer work, right? 

I can't introduce a function call there if there was none in the first place, there is no contract to be enforced -- datasource subclasses were never meant to call a `abstractToRealColumn()` function. So with that in mind I asked myself "how can I still use the `$this->columns[]` array and allow developers to specify a syntax there that maps an abstract type *and* a storage requirement *to* a native type?". That's why I went for a string representation that would just be the type followed by '/' followed by the storage. Does it make sense?

Alternatively I could do both. Introduce a new `abstractToRealColumn()` function that takes `$column`, not just `$type`, and if that method is not found then use the old `$this->columns` array?

markstory replied:
Should we add new abstract types for the various int sizes then? Right now we're doing that implicitly with a less obvious syntax than `'type => 'smallint'`. I recognize that these new types would be less 'abstract' but there is tension between abstract types and fine grained control. Adding explicit types would let us map to existing types when smallint isn't a supported type.

sebastienbarre replied:
Some like byte, integer2? Honestly not sure integer3 is worth something.

markstory replied:
What about `smallint` and `tinyint`?

sebastienbarre replied:
OK. So no need for a `'storage'` field anymore, you would rather have the storage requirements be conveyed by 2 new abstract types, correct?

markstory replied:
I think so. We can map those types to 'integer' in databases that don't support smaller ints like sqlite.

markstory commented on lib/Cake/Test/Case/Model/Datasource/Database/MysqlTest.php:
```diff
@@ -389,6 +389,36 @@ public function testBuildColumn() {
+		$data = array(
+			'name' => 'testName',
+			'type' => 'integer',
+			'storage' => 2
```
How does this new storage option interact with `length`? Instead of tying the sizes to specific bytes, could we do something like:

```php
'my_int' => ['type' => 'integer', 'length' => 4, 'storageType' => 'small']
```

Another option could be `'storageSize' => 2` if byte are better. My concern with using specific bytes, is that we have no error handling for values like '3' or '6', and people will make mistakes that result in an incorrect schema being created. With string names, we can have a lookup of supported values and raise errors on unsupported values.

sebastienbarre replied:
`length` is not an issue. Length was always confusing to MySQL users -- it has absolutely no relation to the storage requirement, it is for display purposes. CakePHP *does* make some assumptions with respect to `tinyint` that have a length of `1` to treat them as an abstract `boolean`, and I made sure this is still working (99% sure). If you had a `tinyint(1)` in your database then you would still end up with an abstract `boolean` in your schema. It's when you are using `tinyint(2)` (or `tinyint(3)`, etc, does not matter the length as long as it's not 1) that you are (still) getting an abstract `integer`. Before my change this `integer` would turn into a 4 bytes `int` in MySQL though. Now it turns back into a `tinyint` as expected, that is, one byte. 

The problem with `'storageType'='small'` is that you are still being too abstract. Who knows how many bytes you mean here, right? If somebody were to implement a new backend, they would still not know how many bytes they can safely use for that column (and not *too many* bytes).

Absolutely, I can rename it to `storageSize`.

In this implementation, if the combination of type and storage is not found, then it defaults to what was found before my change -- in that case, you get an `integer` (4 bytes). But I see what you mean for 6, because `'type' => 'integer', 'storageSize' => 6` should result in something like a `bigint`, right? My gut feeling is this is an "advanced" option, for people who know what they are doing and want tight control of storage because they have *a lot* of rows (we do) -- I would assume these people know that `bigint` exists. Maybe a `honorStorageSize` flag could be used on the datasource itself, and default to false, to bypass my code? Where would be the right place to specify such behavior?

sebastienbarre commented:
@markstory thanks for the feedback I'll have a cleaner version later today.

sebastienbarre committed:
introduce new tinyint, smallint types for smaller storage requirements

sebastienbarre commented:
@markstory alright, here is a first stab at tinyint, smallint. I've tried to update the Postgres, Sqlite and Sqlserver datasource as well, but the Mysql one is the only one I can test properly. Hopefully the continuous integration will reveal issues with the other drivers. Thanks

sebastienbarre committed:
add missing MySQL data type link

sebastienbarre commented:
Tests appear to be passing... which is surprising to me.

markstory commented on lib/Cake/Model/Datasource/Database/Mysql.php:
```diff
@@ -791,6 +785,12 @@ public function column($real) {
 		}
+		if (strpos($col, 'tinyint') !== false) {
+			return 'tinyint';
+		}
```
:+1: this comes after the existing boolean mapping.

sebastienbarre replied:
You bet it does :)

markstory reviewed (approved):
Looks good to me.

markstory commented on lib/Cake/Model/Datasource/Database/Sqlite.php:
```diff
@@ -63,13 +63,17 @@ class Sqlite extends DboSource {
 		'text' => array('name' => 'text'),
 		'integer' => array('name' => 'integer', 'limit' => null, 'formatter' => 'intval'),
+		'smallint' => array('name' => 'integer', 'limit' => null, 'formatter' => 'intval'),
+		'tinyint' => array('name' => 'integer', 'limit' => null, 'formatter' => 'intval'),
```
I took a look at FormHelper and I don't see an existing mappings for `integer` so adding new types shouldn't cause issues there.

chinpei215 replied:
Shouldn't we change [$map](https://github.com/cakephp/cakephp/blob/135a24e0f1f3c842cce7a021d599a18cb77579ff/lib/Cake/View/Helper/FormHelper.php#L1238-L1245) in `FormHelper::_magicOptions`?

sebastienbarre replied:
I guess I should have grepped for `'integer'`. That's a lot of hits, so I just grepped for `'decimal'` instead, assuming that would be code related to datatypes. I'm seeing the following potential candidates for update, would welcome feedback:
* `Console/Command/Task/ModelTask.php` in `fieldValidation()`,
* `Test/Fixture/DatatypeFixture.php`,
* `Test/Fixture/UnsignedFixture.php`,
* and indeed `View/Helper/FormHelper.php`.

the rest is code I touched already under `DataSource` and the corresponding tests.
What do you guys think?

markstory replied:
@chinpei215 Yes, I missed that when I skimmed the code. My mistake.

chinpei215 replied:
Probably `ModelTask` must be updated, as it have to bake numeric validation rules for new smallint/tinyint fields. Regarding the two fixtures, I think we need not to touch it, as they would be fixtures for some test cases of `decimal` fields. Instead, we may have to create `SmallintFixture` and `TinyintFixture` to add new test cases for the new field types.

sebastienbarre replied:
OK I updated `ModelTask` and `FormHelper`. I'm not 100% about the rest. It looks like `DatatypeFixture` should be the right place to add `tiny_int` and `small_int` columns, since `biginteger` and `bool` are already tested there...

sebastienbarre replied:
I update `DatatypeFixture`, let me know if this is satisfactory.

chinpei215 replied:
Sorry, I didn't see the fixtures' name well. Probably, you may have to change *both* fixtures as smallint/tinyint can be declared as `usigned`.

sebastienbarre replied:
Thanks @chinpei215, updating `UnsignedFixture` revealed that I had forgotten to add `smallint` and `tinyint` to `$fieldParameters['unsigned']['types']` in the `Mysql` datasource.

sebastienbarre replied:
We might be all set? :)

chinpei215 replied:
There is need to fix some related test cases using the fixtures.  `MysqlTest::buildColumnUnsignedProvider()` seems like a data provider for such a test case.

sebastienbarre replied:
OK done, thanks. I have no clue who calls `buildColumnUnsignedProvider()`. There is probably some magic going on with the function name.

chinpei215 replied:
See `@dataProvider` annotation in the doc comment of `MysqlTest::testBuildColumnUnsigned()`. That let PHPUnit call the data provider.
https://phpunit.de/manual/current/en/appendixes.annotations.html#appendixes.annotations.dataProvider

sebastienbarre committed:
update ModelTask and FormHelper for tinyint,smallint

sebastienbarre committed:
update more tests, DatatypeFixture

sebastienbarre committed:
fix test for UnsignedFixture

sebastienbarre committed:
more tinyint smallint tests in MySQL

chinpei215 commented:
Looks good to me. Also, I confirmed it with SQL Server on my computer. Now Console/cake schema seems to work with smallint/tinyint fields of that.

sebastienbarre commented:
Hmmm, the only test failing seem to be for PHP 5.4:
```
1) MemcachedEngineTest::testConfigurationConflict

Value was not read %s

Failed asserting that false matches expected 'yay'.
```
I'm a bit suspicious this is related to my change...

chinpei215 commented:
@sebastienbarre  I restarted the job just now. I don't think the build failure relates to your fix. We will see.

sebastienbarre commented:
Thanks. This worked.

chinpei215 commented:
@markstory I received a related report from a user today. He said that `beke fixture` generated `biginteger` fields without default values. In my opinion, [FixtureTask::_generateRecords()](https://github.com/cakephp/cakephp/blob/9536a10d6d048a8024f5bb32548cade2c48ca956/lib/Cake/Console/Command/Task/FixtureTask.php#L338-L341) would relate to the issue, and I think smallint/tinyint types also would cause the same issue. I can fix the issue after this commit get merged, but I would like to note it here, as I think this ticket will be a good example when someone wants to add new types to 2.x branch.

markstory commented:
I'll get this merged once I get a pull request for the 3.x version opened.

sebastienbarre commented:
@markstory I'm a bit confused. I'm never used 3.x. Who is porting that work on 3.x?

markstory commented:
@sebastienbarre I am.

ravage84 commented on lib/Cake/Model/Datasource/Database/Mysql.php:
```diff
@@ -109,6 +113,8 @@ class Mysql extends DboSource {
 /**
  * MySQL column definition
  *
+ * @link https://dev.mysql.com/doc/refman/5.7/en/data-types.html MySQL Data Types
```
Move this after the ``@var``

ravage84 commented on lib/Cake/Model/Datasource/Database/Postgres.php:
```diff
@@ -51,13 +51,17 @@ class Postgres extends DboSource {
 /**
  * Columns
  *
+ * @link https://www.postgresql.org/docs/9.6/static/datatype.html PostgreSQL Data Types
```
Move this after the ``@var``

ravage84 commented on lib/Cake/Model/Datasource/Database/Sqlite.php:
```diff
@@ -63,13 +63,17 @@ class Sqlite extends DboSource {
 /**
  * SQLite3 column definition
  *
+ * @link https://www.sqlite.org/datatype3.html Datatypes In SQLite Version 3
```
Move this after the ``@var``

dereuromark replied:
Should be done by a sniffer later. Now Just unnecessary Work and waste of time IMO.

ravage84 commented on lib/Cake/Model/Datasource/Database/Sqlserver.php:
```diff
@@ -84,13 +84,17 @@ class Sqlserver extends DboSource {
 /**
  * MS SQL column definition
  *
+ * @link https://msdn.microsoft.com/en-us/library/ms187752.aspx SQL Server Data Types
```
Move this after the ``@var``

ravage84 commented on lib/Cake/Test/Case/Model/Datasource/Database/MysqlTest.php:
```diff
@@ -3368,6 +3376,26 @@ public function buildColumnUnsignedProvider() {
 				),
 				'`testName` decimal UNSIGNED DEFAULT 1'
+			),
+			//set #9
```
Instead of comments, you can use the array keys for documenting the data set.
The advantage is that it would show the name of the data set in case of an error in the error message.
It might even show it in the coverage, not sure though.

chinpei215 replied:
But it has also some disadvantage. Even if people uses the same key, PHP doesn't complain it. Some important data might be not provided in such cases. If our code sniffer cannot complain it, I am not in favor of that.

ravage84 reviewed (commented):
Minor nitpicking from me, as usual 😼

sebastienbarre committed:
coding/doc standard, move @link after @var, per req

sebastienbarre commented:
@ravage84 I moved the @link. As far as "//set #9" is concerned, it's a bit above my head, so I'll let you guys take care of it if needed?

markstory committed:
Rename (small|tiny)int to (small|tiny)integer

Make the new types consistent with the biginteger.

markstory committed:
Add default value generation for all integer types.

This resolves the bigint issue raised by @chinpei in 10347 and adds
default values for the new small & tiny integer types.

sebastienbarre committed:
Merge pull request #1 from cakephp/issue-10347

A few consistency fixes for small/tiny integer support
